### ISSUE TITLE
Messages get dropped if sent before client's resume_timeout 

### ISSUE THREAD

**Issue by amitojc** [processone/ejabberd#1160]
Hi,

I've recently compiled latest ejabberd code from github on my Arch linux machine. I'm new to ejabberd so I may have made a newbie mistake.

I'm facing an issue where in a particular scenario packets do not reach the target client. This happens if the message is sent __after__ client gets disconnected but __before__ client's session is closed due to resume_timeout.

I've followed instructions as mentioned over here https://www.ejabberd.im/faq/tcp , but still I've been unable to handle all cases.

The relevant parts of my ejabberd.yml look like

```
listen:
  - 
    port: 5222
    module: ejabberd_c2s
    max_stanza_size: 65536
    resend_on_timeout: true
    resume_timeout: 60
    shaper: c2s_shaper
    access: c2s

  - 
    ## Not sure if resend_on_timeout is required here
    port: 5280
    module: ejabberd_http
    resend_on_timeout: true
    resume_timeout: 60
    http_bind: true

modules:
  ...
  mod_offline:
    access_max_user_messages: max_user_offline_messages
  mod_ping:
    send_pings: true
    ping_interval: 60
    timeout_action: kill
```

Here's the output of ejabberd log corresponding to the client activities:

- Client A disconnects uncleanly
```
2016-06-19 10:59:51.369 [info] <0.549.0>@ejabberd_c2s:fsm_next_state:2454 Waiting for resumption of stream for 9739937980@gappi.in/Smack
```
- Client B Logs in and sends a message to A
```
2016-06-19 11:00:00.320 [info] <0.510.0>@ejabberd_listener:accept:333 (#Port<0.19876>) Accepted connection 139.162.34.247:41186 -> 139.162.34.247:5222
2016-06-19 11:00:00.575 [info] <0.559.0>@ejabberd_c2s:wait_for_sasl_response:900 ({socket_state,fast_tls,{tlssock,#Port<0.19876>,#Port<0.19877>},<0.558.0>}) Accepted authentication for echo by undefined from 137
2016-06-19 11:00:00.715 [info] <0.559.0>@ejabberd_c2s:open_session:1111 ({socket_state,fast_tls,{tlssock,#Port<0.19876>,#Port<0.19877>},<0.558.0>}) Opened session for echo@gappi.in/13353511976792329881266
2016-06-19 11:00:00.776 [info] <0.559.0>@ejabberd_c2s:handle_enable:2700 Stream management with resumption enabled for echo@gappi.in/13353511976792329881266
2016-06-19 11:00:00.984 [info] <0.559.0>@ejabberd_c2s:terminate:1778 ({socket_state,fast_tls,{tlssock,#Port<0.19876>,#Port<0.19877>},<0.558.0>}) Close session for echo@gappi.in/13353511976792329881266
```
- Client A session is finally closed due to resume_timeout
```
2016-06-19 11:00:51.370 [info] <0.549.0>@ejabberd_c2s:terminate:1778 ({socket_state,gen_tcp,#Port<0.19869>,<0.548.0>}) Close session for 9739937980@gappi.in/Smack
```
- Client A logs back in
```
2016-06-19 11:01:14.078 [info] <0.510.0>@ejabberd_listener:accept:333 (#Port<0.19879>) Accepted connection 122.172.241.21:26597 -> 139.162.34.247:5222
2016-06-19 11:01:14.707 [info] <0.566.0>@ejabberd_c2s:wait_for_sasl_response:900 ({socket_state,gen_tcp,#Port<0.19879>,<0.565.0>}) Accepted authentication for 9739937980 by undefined from 122.172.241.21
2016-06-19 11:01:14.868 [info] <0.566.0>@ejabberd_c2s:open_session:1111 ({socket_state,gen_tcp,#Port<0.19879>,<0.565.0>}) Opened session for 9739937980@gappi.in/Smack
2016-06-19 11:01:14.966 [info] <0.566.0>@ejabberd_c2s:handle_enable:2700 Stream management with resumption enabled for 9739937980@gappi.in/Smack
```

As per this guide, https://www.ejabberd.im/faq/tcp, after the resume_timeout expires, the unacked messages should have been handled by mod_offline. But this is not happening. 

Result : No message reaches client.

Any pointers?

**Comment by weiss** [comment]
I'd set `loglevel: 5` or use a module such as [`mod_logxml`][1] to trace the missing message(s) and any error responses.  If possible, I'd also dump the stream with Smack.

And once this issue is tracked down, I'd set `resend_on_timeout: if_offline` unless you *know* no user will run multiple clients in parallel.

[1]: https://github.com/processone/ejabberd-contrib/tree/master/mod_logxml

**Comment by amitojc** [comment]
Thank you for your response. I think I've narrowed down the problem.

I uninstalled the ejabberd that I had compiled(from Github) and instead installed the version available in Arch Linux's repository - https://www.archlinux.org/packages/community/i686/ejabberd/.

Now the resend_on_timeout works perfectly. So I think the problem was due to some of the recent commits to ejabberd which introduced a bug, or I had made some mistake while compiling ejabberd. 

Moreover, self compiled installation, I could not view the 'offline' messages in the web admin. Mod_offline worked, and web admin would show that a user had 1 offline message. But on clicking the offline messages link, I used to get "NOT FOUND" page. That problem has also gotten resolved by installing ejabberd  from Arch Linux's repository.

So maybe these two problems were related, but I have no idea about ejabberd's code.

**Comment by weiss** [comment]
This is now fixed in Git, thanks for the report.

**Comment by amitojc** [comment]
Wow, I reported an actual bug! All this time I was thinking that problem was at my side. 

However, I'm getting another warning "1 stanzas were not acknowledged by 9739937980@gappi.in/Smack" when I resume a connection. I'll open a new issue for that. Thanks.

**Comment by weiss** [comment]
I forgot to mention that I fixed the stream management issue but not the web interface thing, as I wasn't able to reproduce that.  If you run into that one again, please file a new issue.

**Comment by amitojc** [comment]
@weiss  - I'm facing the same issue again. Is there a chance that this fix has been mistakenly reverted in 16.12-beta1?

**Comment by weiss** [comment]
The fix wasn't reverted, but I guess this was clarified in #1415.

**Comment by amitojc** [comment]
Yea. I didn't read the documentation properly. Thanks again.
